File: leetCode/C++/268-missingNumber.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int expected_sum = ((nums.size() + 1) * nums.size()) / 2;
        int real_sum = 0;
        for (const auto &num: nums)
            real_sum += num;
        return expected_sum - real_sum;
    }
};

int main(void) {
    Solution s;
    vector<int> nums {0, 1, 2, 3, 4, 5, 7};
    cout << s.missingNumber(nums) << endl;
    return 0;
}

File: leetCode/Python/268-missingNumber.py
# Time:  O(n)
# Space: O(1)
#
# Given an array containing n distinct numbers taken from
# 0, 1, 2, ..., n, find the one that is missing from the array.
#
# For example,
# Given nums = [0, 1, 3] return 2.
#
# Note:
# Your algorithm should run in linear runtime complexity.
# Could you implement it using only constant extra space complexity?
#

class Solution1(object):
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return reduce(operator.xor, nums, \
                      reduce(operator.xor, xrange(len(nums) + 1)))

class Solution(object):
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums) + 1
        summation = (length - 1) * length / 2
        return summation - sum(nums)

if __name__ == '__main__':
   s = Solution()
   print s.missingNumber([0])
   print s.missingNumber([0, 1])
   print s.missingNumber([0, 2])

